pennylane_qiskit/qiskit_device.py

```diff
@@ -114,31 +94,44 @@ class QiskitDevice(Device, abc.ABC):
     _state_backends = {"statevector_simulator", "unitary_simulator"}
     """set[str]: Set of backend names that define the backends
     that support returning the underlying quantum statevector"""
+    _hw_backends = {"qasm_simulator", "ibmq_qasm_simulator"}
+    """set[str]: Set of backend names that define the backends that support
+    hardware or hardware simulator execution and storing samples in memory."""
 
     operations = set(_operation_map.keys())
     observables = {"PauliX", "PauliY", "PauliZ", "Identity", "Hadamard", "Hermitian"}
 
     hw_analytic_warning_message = (
-        "The analytic calculation of expectations and variances "
-        "is only supported on statevector backends, not on the {}. "
-        "The obtained result is based on sampling."
+        "The analytic calculation of expectations, variances and "
+        "probabilities is only supported on statevector backends, not on the {}. "
+        "Such statistics obtained from this device are estimates based on samples."
     )
 
     _eigs = {}
 
     def __init__(self, wires, provider, backend, shots=1024, **kwargs):
         super().__init__(wires=wires, shots=shots)
 
+        # Keep track if the user specified analytic to be True
+        analytic_was_specified = "analytic" in kwargs and kwargs["analytic"]
         self.analytic = kwargs.pop("analytic", False)
 
+        if self.analytic and backend not in self._state_backends:
+
+            if analytic_was_specified:
```

Comment:
How come this if statement is required? Shouldn't the warning be raised if `self.analytic == True`, regardless if the user specified it or not?

Reply:
Unfortunately, we are setting `analytic=True` right now in PennyLane by default: https://github.com/XanaduAI/pennylane/blob/8a5a164b6b83fd54a141bae5f604c32ecbaf9ba7/pennylane/_qubit_device.py#L72

We could fix it quickly, but I'd imagine that it would be better to overall rethink when and for which devices we define the `analytic` attribute in the first place. E.g. if we go with subclassing `QubitDevice`, then the `StateVectorDevice` device could define `analytic` whereas `HardwareDevice` would not. Using inheritance would overall allow us to sort out this part of the logic just by using Python itself.

Reply:
👍